runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/ExecutableStage.java

```diff
@@ -84,64 +81,72 @@
    * follows:
    *
    * <ul>
-   *   <li>The {@link PTransform#getSubtransformsList()} contains the result of {@link
-   *       #getTransforms()}.
+   *   <li>The {@link PTransform#getSubtransformsList()} contains no subtransforms. This ensures
+   *       that executable stages are treated as primitive transforms.
    *   <li>The only {@link PCollection} in the {@link PTransform#getInputsMap()} is the result of
    *       {@link #getInputPCollection()}.
    *   <li>The output {@link PCollection PCollections} in the values of {@link
    *       PTransform#getOutputsMap()} are the {@link PCollectionNode PCollections} returned by
    *       {@link #getOutputPCollections()}.
+   *   <li>The {@link FunctionSpec} contains an {@link ExecutableStagePayload} which has its input
+   *       and output PCollections set to the same values as the outer PTransform itself. It further
+   *       contains the environment set of transforms for this stage.
    * </ul>
+   *
+   * <p>The executable stage can be reconstructed from the resulting {@link ExecutableStagePayload}
+   * and components alone via {@link #fromPayload(ExecutableStagePayload, Components)}.
    */
   default PTransform toPTransform() {
+    ExecutableStagePayload.Builder payload = ExecutableStagePayload.newBuilder();
+
+    payload.setEnvironment(getEnvironment());
+
+    PCollectionNode input = getInputPCollection();
+    payload.setInput(input.getId());
+
+    for (PTransformNode transform : getTransforms()) {
+      payload.addTransforms(transform.getId());
```

Comment:
I'm not huge on the double iteration, given that I think you can populate both the payload and the transform builder in the same places to maximize mental locality (e.g. populate `payload.setInput` and then `pt.putInput("input", `)